pkg/pillar/zedcloud/authen.go

```diff
@@ -0,0 +1,508 @@
+)
+
+// given an envelope protobuf received from controller, verify the authentication
+func verifyAuthentication(c []byte, skipVerify bool) ([]byte, types.HTTPExtraStatus, error) {
+	extraSt := types.ExtStatusNone
+	sm := &zauth.AuthContainer{}
+	err := proto.Unmarshal(c, sm)
+	if err != nil {
+		log.Infof("verifyAuthentication: can not unmarshal authen content, %v\n", err)
+		return nil, extraSt, err
+	}
+
+	data := sm.AuthPayload.GetPayload()
+	if !skipVerify { // no verify for /certs itself
+		if len(sm.GetSenderCertHash()) != hashSha256Len16 &&
+			len(sm.GetSenderCertHash()) != hashSha256Len32 {
+			log.Infof("verifyAuthentication: senderCertHash length %d\n", len(sm.GetSenderCertHash()))
+			err := fmt.Errorf("verifyAuthentication: senderCertHash length error")
+			return nil, types.ExtStatusHashSizeError, err
+		}
+
+		if sm.Algo != zauth.HashAlgorithm_HASH_SHA256_16bytes &&
+			sm.Algo != zauth.HashAlgorithm_HASH_SHA256_32bytes {
+			log.Infof("verifyAuthentication: hash algorithm is not supported\n")
+			err := fmt.Errorf("verifyAuthentication: hash algorithm is not supported")
+			return nil, types.ExtStatusAlgoFail, err
+		}
+
+		cert, err := getCloudCert()
+		if err != nil {
+			log.Infof("verifyAuthentication: can not get server cert, %v\n", err)
+			return nil, extraSt, err
+		}
+
+		if sm.Algo == zauth.HashAlgorithm_HASH_SHA256_32bytes &&
+			bytes.Compare(sm.GetSenderCertHash(), cloudCertHash) != 0 ||
+			sm.Algo == zauth.HashAlgorithm_HASH_SHA256_16bytes &&
+				bytes.Compare(sm.GetSenderCertHash(), cloudCertHash[:hashSha256Len16]) != 0 {
+			err := fmt.Errorf("verifyAuthentication: local server cert hash does not match in authen")
+			log.Infof("verifyAuthentication: local server cert hash(%d) does not match in authen (%d) %v, %v",
+				len(cloudCertHash), len(sm.GetSenderCertHash()), cloudCertHash, sm.GetSenderCertHash())
+			return nil, types.ExtStatusCertMiss, err
+		}
+
+		hash := ComputeSha(data)
+		err = verifyAuthSig(sm.GetSignatureHash(), cert, hash)
+		if err != nil {
+			log.Infof("verifyAuthentication: verifyAuthSig error %v\n", err)
+			return nil, types.ExtStatusSignVerifyFail, err
+		}
+		log.Infof("verifyAuthentication: ok\n")
+	}
+	return data, extraSt, nil
+}
+
+func getCloudCert() (*x509.Certificate, error) {
+	var sCert *x509.Certificate
+	if serverLeafCert == nil {
+		certBytes, err := ioutil.ReadFile(types.ServerCertName)
+		if err != nil {
+			log.Infof("getCloudCert: can not read in server cert file, %v\n", err)
+			return sCert, err
+		}
+		block, _ := pem.Decode(certBytes)
+		if block == nil {
+			err := fmt.Errorf("getCloudCert: can not get client Cert")
+			return sCert, err
+		}
+
+		sCert, err = x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			log.Infof("getCloudCert: can not parse cert %v\n", err)
+			return sCert, err
+		}
+
+		// hash verify using PEM bytes from cloud
+		cloudCertHash = ComputeSha(certBytes)
+		serverLeafCert = sCert
+
+	} else {
+		sCert = serverLeafCert
```

Comment:
So after a fetch of a new server cert due to a miss this code will continue to use the server cert it read from the file on startup? In the next sprint we need to add some Update function which updates the zedcloudContext with the new server cert.

Reply:
Good point. missed the update in this case.